pkg/apis/core/validation/validation.go

```diff
@@ -4217,11 +4201,8 @@ func ValidateServiceUpdate(service, oldService *core.Service) field.ErrorList {
 		if oldService.Spec.Type != core.ServiceTypeExternalName && oldService.Spec.ClusterIP != "" {
 			allErrs = append(allErrs, ValidateImmutableField(service.Spec.ClusterIP, oldService.Spec.ClusterIP, field.NewPath("spec", "clusterIP"))...)
 		}
-		// notes:
-		// we drop the IPFamily field when the Dualstack gate is off.
-		// once the gate is on, we start assigning default ipfamily according to cluster settings. in other words
-		// though the field is immutable, we allow (onetime) change from nil==> to value
-		if oldService.Spec.IPFamily != nil {
+		// a user is not allowed to change the IPFamily field, but they are allowed to nil it out (old clients that don't know about IPFamily)
+		if oldService.Spec.IPFamily != nil && service.Spec.IPFamily != nil {
```

Comment:
rather than relax validation here, I would have expected logic in PrepareForUpdate like:
```
if oldService.Spec.IPFamily != nil {
  if newService.Spec.IPFamily == nil {
    // old clients might drop the unknown IPFamily field, so preserve it
    family := *oldService.Spec.IPFamily
    newService.Spec.IPFamily = &family
  }
}
```

Reply:
I guess the alternative is letting them nil it out, then re-defaulting in https://github.com/kubernetes/kubernetes/blob/c9003a268dff6700506929b247847341ea4d5b33/pkg/apis/core/v1/defaults.go#L136-L160

which do we prefer?

Reply:
we need it in both places I think - what if you changed config from dual stack to ipv6 only and go delete all you v4 services?

Reply:
After our discussion restructured the method, ensured defaulting is in Before* and also have the "bypass when unchanged" in both places (validation and before*)